webui/react/src/stores/users.tsx

```diff
@@ -13,230 +14,108 @@ type UsersPagination = {
+      this.#currentUserId.set(Loaded(response.id));
     } catch (e) {
-      handleError(e);
+      handleError(e, { publicSubject: 'Unable to fetch current user.' });
     }
-  }, [canceler, updateCurrentUser, currentUser, updateUsers]);
-};
+  };
 
-export const useEnsureUsersFetched = (canceler: AbortController): (() => Promise<void>) => {
-  const context = useContext(UsersContext);
-
-  if (context === null) {
-    throw new Error('Attempted to use useFetchUsers outside of Users Context');
-  }
-
-  const { updateUsers, updateUsersByKey, usersByKey } = context;
-
-  return useCallback(
-    async (cfg?: FetchUsersConfig): Promise<void> => {
-      const config = cfg ?? {};
-      const usersPagination = usersByKey.get(encodeParams(config));
-
-      if (usersPagination) return;
-
-      try {
-        const response = await getUsers(config, { signal: canceler.signal });
-        const usersPages = {
-          pagination: response.pagination,
-          users: response.users.map((user) => user.id),
-        };
-
-        updateUsersByKey((prevState) => prevState.set(encodeParams(config), usersPages));
-        updateUsers((prevState) => {
-          return prevState.withMutations((map) => {
-            response.users.forEach((user) => map.set(user.id, user));
-          });
-        });
-      } catch (e) {
-        handleError(e);
-      }
-    },
-    [canceler, updateUsers, usersByKey, updateUsersByKey],
-  );
-};
+  public updateCurrentUser = (id: number | null) => {
+    if (id === null) this.#currentUserId.set(NotLoaded);
+    else this.#currentUserId.set(Loaded(id));
+  };
 
-export const useUsers = (cfg?: FetchUsersConfig): Loadable<Readonly<UserPage>> => {
-  const context = useContext(UsersContext);
+  public getUsers = (cfg?: FetchUsersConfig): Observable<Loadable<DetailedUserList>> => {
+    const config = cfg ?? {};
 
-  if (context === null) {
-    throw new Error('Attempted to use useUsers outside of Users Context');
-  }
-  const config = cfg ?? {};
-  const usersPagination = context.usersByKey.get(encodeParams(config));
+    return this.#usersByKey.select((map) => {
+      const usersPagination = map.get(encodeParams(config));
 
-  if (!usersPagination) return NotLoaded;
+      if (!usersPagination) return NotLoaded;
 
-  const userPage: UserPage = {
-    pagination: usersPagination.pagination,
-    users: usersPagination.users.flatMap((userId) => {
-      const user = context.users.get(userId);
+      const userPage: DetailedUserList = {
+        pagination: usersPagination.pagination,
+        users: usersPagination.users.flatMap((userId) => {
+          const user = this.#users.get().get(userId);
 
-      return user ? [user] : [];
-    }),
+          return user ? [user] : [];
+        }),
+      };
+
+      return Loaded(userPage);
+    });
   };
 
-  return Loaded(userPage);
-};
+  public ensureUsersFetched = async (
+    canceler: AbortController,
+    cfg?: FetchUsersConfig,
+  ): Promise<void> => {
```

Comment:
My one note is that in idiomatic reactive programming, _actions_ (fetch function in the case of stores) should always return `void`. Exposing the promise here allows consumers to await the promise when they should instead be subscribing to the observables returned by the various `get` functions.

Reply:
That's a good point, I'll go remove those promises.